mp2-common/src/storage_key.rs

```diff
@@ -28,42 +40,77 @@ const INPUT_TUPLE_LEN: usize = 2 * INPUT_ELEMENT_LEN;
 #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
-/// Wires associated with the MPT key from the keccak computation of location
 pub struct KeccakMPTWires {
-    /// Actual keccak wires created for the computation of the "location" for
-    /// the storage slot
-    pub keccak_location: ByteKeccakWires<INPUT_PADDED_LEN>,
-    /// Actual keccak wires created for the computation of the final MPT key
-    /// from the location. THIS is the one to use to look up a key in the
+    /// Actual Keccak wires created for the computation of the base for the storage slot
+    pub keccak_base: ByteKeccakWires<INPUT_PADDED_LEN>,
+    /// Actual Keccak wires created for the computation of the final MPT key
+    /// from the location. this is the one to use to look up a key in the
     /// associated MPT trie.
     pub keccak_mpt_key: KeccakWires<{ PAD_LEN(HASH_LEN) }>,
-    /// The MPT key derived in circuit from the storage slot, in NIBBLES
+    /// The MPT key derived in circuit from the storage slot in nibbles
     /// TODO: it represents the same information as "exp" but in nibbles.
     /// It doesn't need to be assigned, but is used in the higher level circuits
     pub mpt_key: MPTKeyWire,
 }
 
+// The Keccak MPT computation
+// It's only used for mapping slot.
 struct KeccakMPT;
 
 impl KeccakMPT {
+    /// Build the Keccak MPT with no offset (offset = 0).
     fn build<F: RichField + Extendable<D>, const D: usize>(
         b: &mut CircuitBuilder<F, D>,
         inputs: VectorWire<Target, INPUT_PADDED_LEN>,
     ) -> KeccakMPTWires {
-        let keccak_location = KeccakCircuit::<{ INPUT_PADDED_LEN }>::hash_to_bytes(b, &inputs);
-        // keccak(location) - take the output and copy it in a slice large
-        // enough for padding.
-        let mut padded_location = [b.zero(); PAD_LEN(HASH_LEN)];
-        padded_location[0..HASH_LEN].copy_from_slice(&keccak_location.output.arr);
+        let keccak_base = KeccakCircuit::<{ INPUT_PADDED_LEN }>::hash_to_bytes(b, &inputs);
+        let location = keccak_base.output.arr;
+
+        Self::build_location(b, keccak_base, location)
+    }
+
+    /// Build the Keccak MPT with a specified offset of Uint32.
+    fn build_with_offset<F: SerializableRichField<D> + Extendable<D>, const D: usize>(
+        b: &mut CircuitBuilder<F, D>,
+        inputs: VectorWire<Target, INPUT_PADDED_LEN>,
+        offset: Target,
+    ) -> KeccakMPTWires {
+        // location = keccak(inputs) + offset
+        let keccak_base = KeccakCircuit::<{ INPUT_PADDED_LEN }>::hash_to_bytes(b, &inputs);
+        let base = keccak_base.output.arr.pack(b, Endianness::Big);
+        let base = UInt256Target::new_from_be_target_limbs(&base).unwrap();
+        let offset = UInt256Target::new_from_target_unsafe(b, offset);
+        let (location, overflow) = b.add_u256(&base, &offset);
+        b.assert_zero(overflow.0);
+        let location = unpack_u32s_to_u8_targets(b, location.to_targets(), Endianness::Big)
```

Comment:
It could be more efficient if we don't have to range check each byte of the uncompressed witness. I think that it should be ok in this case to not range-check since we are decomposing just to hash bytes with Keccak, so even if the decomposition is not correct, as soon as it can be packed in the same `u32' limbs, then the results of Keccak should be unaffected. Wdyt?

Otherwise, if we need to range-check, then it would be more efficient only if we introduce a further lookup-table for byte range-checks (which could be useful also elsewhere, though we discussed that it will have a negligible impact over the cost of Keccak).